java/runtime/src/main/java/io/ray/runtime/io/MemoryBuffer.java

```diff
@@ -0,0 +1,968 @@
+package io.ray.runtime.io;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.base.Preconditions;
+import java.nio.BufferOverflowException;
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.ReadOnlyBufferException;
+
+/**
+ * A class for operations on memory managed by Ray. The buffer may be backed by heap memory (byte
+ * array) or by off-heap memory. If the memory is off-heap memory, it will be the memory allocated
+ * by ray object store.
+ *
+ * <p>This class is based on org.apache.flink.core.memory.MemorySegment and
+ * org.apache.arrow.memory.ArrowBuf, we add this class mainly for:
+ *
+ * <ul>
+ *   <li>read/write data into a chunk of direct memory.
+ *   <li>additional binary compare, swap, and copy methods.
+ *   <li>little-endian access.
+ *   <li>independent read/write index.
+ * </ul>
+ *
+ * <p>Note that this class is designed to final so that all the methods in this class can be inlined
+ * by the just-in-time compiler.
+ */
+public final class MemoryBuffer {
+  // The unsafe handle for transparent memory copied (heap / off-heap).
+  private static final sun.misc.Unsafe UNSAFE = Platform.UNSAFE;
+  // The beginning of the byte array contents, relative to the byte array object.
+  private static final long BYTE_ARRAY_BASE_OFFSET = UNSAFE.arrayBaseOffset(byte[].class);
+  // Constant that flags the byte order. Because this is a boolean constant, the JIT compiler can
+  // use this well to aggressively eliminate the non-applicable code paths.
+  private static final boolean LITTLE_ENDIAN = (ByteOrder.nativeOrder() == ByteOrder.LITTLE_ENDIAN);
+
+  // The heap byte array object relative to which we access the memory.
+  // Is non-null if the memory is on the heap, and is null, if the memory is off the heap.
+  // If we have this buffer, we must never void this reference, or the memory buffer will point
+  // to undefined addresses outside the heap and may in out-of-order execution cases cause
+  // buffer faults.
+  private byte[] heapMemory;
+  // The direct byte buffer that allocated the off-heap memory. This memory buffer holds a
+  // reference to that buffer, so as long as this memory buffer lives, the memory will not be
+  // released.
+  private ByteBuffer offHeapBuffer;
+  // The readable/writeable range is [address, addressLimit).
+  // The address to the data, relative to the heap memory byte array. If the heap memory byte array
+  // is null, this becomes an absolute memory address outside the heap.
```

Comment:
```suggestion
  // If the data in on the heap, this is the relative offset to the `heapMemory` byte array. 
  // If the data is off the heap, this is the absolute memory address.
```

Reply:
Maybe also change the comments of the other fields to this format. It would be more clear.

Reply:
Good suggestion, will rename all related fields comments